nixos/systemd-repart: enable creating root partition

nikstur opened the pull request:
###### Description of changes

Add the ability to create the root partition while in the initrd. This enables to build system image with only a Nix store where the root partition is dynamically created.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nikstur committed:
nixos/systemd-repart: enable creating root partition

arianvp commented on nixos/modules/system/boot/systemd/repart.nix:
```diff
@@ -98,16 +113,31 @@ in
+          let
+            # Imitates `systemd-escape --path` but will not work correctly for
+            # corner cases.
+            escapePath = path: lib.replaceStrings [ "/" ] [ "-" ] (lib.removePrefix "/" path);
```
We have an escape-path utility in systemd-lib.nix

nikstur replied:
Do you mean [mkPathSafeName](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/systemd-lib.nix#L13)? I thought about using it but it doesn't really do what I need (i.e strip the first `/` and change `/` in the middle to `-`). Is it supposed to mimic `systemd-escape --path`? Then maybe it would make sense to change it. However I don't really know the blast radius of messing up this escape function.

ElvishJerricco replied:
@nikstur He means `escapeSystemdPath`, right?

nikstur replied:
Awesome, I just looked in the wrong places.

ElvishJerricco reviewed (commented):
My only concerns are the `escapeSystemdPath` thing and another thing I just realized. We definitely don't want it looking at the config files in `/sysroot/etc/repart.d` before activation because those could be out of date. So the `after = ["sysroot.mount"];` should almost certainly be `after = ["nixos-activation.service"];` That should also be added to requires.

Also, FYI, it's generally a little more convenient / appropriate to use `systemd.services.foo.after` rather than `systemd.services.foo.unitConfig.After` (same goes for `requires`). This makes it easier to add to with the nixos module system. Not a big deal though.

nikstur commented:
> Also, FYI, it's generally a little more convenient / appropriate to use `systemd.services.foo.after` rather than `systemd.services.foo.unitConfig.After` (same goes for `requires`). This makes it easier to add to with the nixos module system. Not a big deal though.

Good to know, changed it accordingly.

> We definitely don't want it looking at the config files in `/sysroot/etc/repart.d` before activation because those could be out of date.

This is not an issue in the initrd since the config files are now baked into the initrd and we tell `systemd-repart` explicitly to look for the configs in the initrd with `--definitions`. It will not try to look for them elsewhere.

However in userspace, this might be an issue. I added the ordering there.

nikstur commented on nixos/modules/system/boot/systemd/repart.nix:
```diff
@@ -119,6 +145,11 @@ in
+      services.systemd-repart = {
+        requires = [ "nixos-activation.service" ];
+        after = [ "nixos-activation.service" ];
+      };
```
Ok now I see how this doesn't make sense here :D

nikstur replied:
Removed it.

nikstur commented:
@arianvp @ElvishJerricco is there anything else to be done? Or is this good to go?
